obstore/python/obstore/auth/azure.py

```diff
@@ -0,0 +1,200 @@
+"""Credential providers for Azure Cloud Storage that use [`azure.identity`][].
+
+[`azure.identity`]: https://learn.microsoft.com/en-us/python/api/overview/azure/identity-readme
+"""
+
+from __future__ import annotations
+
+from datetime import UTC, datetime, timedelta
+from typing import TYPE_CHECKING
+
+import azure.identity
+import azure.identity.aio
+
+if TYPE_CHECKING:
+    from obstore.store import AzureCredential
+
+    AzureCredentialUnionType = (
+        azure.identity.AuthorizationCodeCredential
+        | azure.identity.AzureCliCredential
+        | azure.identity.AzureDeveloperCliCredential
+        | azure.identity.AzurePipelinesCredential
+        | azure.identity.AzurePowerShellCredential
+        | azure.identity.CertificateCredential
+        | azure.identity.ChainedTokenCredential
+        | azure.identity.ClientAssertionCredential
+        | azure.identity.ClientSecretCredential
+        | azure.identity.DefaultAzureCredential
+        | azure.identity.DeviceCodeCredential
+        | azure.identity.EnvironmentCredential
+        | azure.identity.InteractiveBrowserCredential
+        | azure.identity.ManagedIdentityCredential
+        | azure.identity.OnBehalfOfCredential
+        | azure.identity.SharedTokenCacheCredential
+        | azure.identity.UsernamePasswordCredential
+        | azure.identity.VisualStudioCodeCredential
+        | azure.identity.WorkloadIdentityCredential
+    )
+
+    AzureAsyncCredentialUnionType = (
+        azure.identity.aio.AuthorizationCodeCredential
+        | azure.identity.aio.AzureCliCredential
+        | azure.identity.aio.AzureDeveloperCliCredential
+        | azure.identity.aio.AzurePipelinesCredential
+        | azure.identity.aio.AzurePowerShellCredential
+        | azure.identity.aio.CertificateCredential
+        | azure.identity.aio.ChainedTokenCredential
+        | azure.identity.aio.ClientAssertionCredential
+        | azure.identity.aio.ClientSecretCredential
+        | azure.identity.aio.DefaultAzureCredential
+        | azure.identity.aio.EnvironmentCredential
+        | azure.identity.aio.ManagedIdentityCredential
+        | azure.identity.aio.OnBehalfOfCredential
+        | azure.identity.aio.SharedTokenCacheCredential
+        | azure.identity.aio.VisualStudioCodeCredential
+        | azure.identity.aio.WorkloadIdentityCredential
+    )
+
+
+class AzureAuthCredentialProvider:
```

Comment:
I named it `GoogleAuthCredentialProvider` because the module is `google.auth`. So this would be `AzureIdentityCredentialProvider` because the module used is `azure.identity`. That said, that name is super long so I'm open to alternate suggestions.

Reply:
How do you feel about `AzureCredentialProvider`?

Reply:
That's fine; maybe we should rename the `GoogleAuth` too

Reply:
Done. (For Google too.)